File: apps/ai/app/agents/fingerprint.py
"""Solution fingerprinting — Python side.

Byte-identical algorithm to packages/db/prisma/seed/fingerprint.ts. The seed
computes the reference solution's fingerprint at publish time; the Response
Guard computes the model's proposed code the same way and compares.

IMPORTANT: if you change one, change both. tests/test_guard/test_fingerprint.py
holds shared fixtures precisely so drift is caught rather than silently
disabling leak detection.
"""

from __future__ import annotations

import re
from collections import Counter

KEYWORDS: set[str] = {
    # control flow
    "if", "else", "elif", "for", "while", "do", "switch", "case", "default",
    "break", "continue", "return", "yield", "goto",
    # declarations
    "def", "function", "fn", "func", "lambda", "class", "struct", "interface",
    "enum", "const", "let", "var", "val", "static", "public", "private",
    "protected", "void", "new", "delete", "this", "self", "super", "extends",
    "implements", "import", "from", "include", "using", "namespace", "package",
    # types
    "int", "long", "short", "char", "float", "double", "bool", "boolean",
    "string", "str", "list", "dict", "set", "map", "vector", "array", "auto",
    "unsigned", "signed", "size_t",
    # values / word operators
    "true", "false", "null", "none", "nil", "nullptr", "undefined",
    "and", "or", "not", "in", "is", "try", "catch", "except", "finally",
    "throw", "raise", "with", "as", "pass", "global", "nonlocal",
    # library calls whose presence is structural rather than cosmetic
    "range", "len", "enumerate", "sorted", "sort", "append", "push", "pop",
    "min", "max", "sum", "abs", "print", "push_back", "insert", "get",
}

_IDENT_START = re.compile(r"[A-Za-z_]")
_IDENT_PART = re.compile(r"[A-Za-z0-9_]")
_NUMBER_PART = re.compile(r"[0-9.eExXaAbBcCdDfF]")


def _strip_literals_and_comments(src: str) -> str:
    out: list[str] = []
    i = 0
    n = len(src)

    while i < n:
        ch = src[i]
        nxt = src[i + 1] if i + 1 < n else ""

        if (ch == "/" and nxt == "/") or ch == "#":
            while i < n and src[i] != "\n":
                i += 1
            continue

        if ch == "/" and nxt == "*":
            i += 2
            while i + 1 < n and not (src[i] == "*" and src[i + 1] == "/"):
                i += 1
            i += 2
            continue

        if ch in {'"', "'"} and src[i : i + 3] == ch * 3:
            quote = ch * 3
            i += 3
            end = src.find(quote, i)
            i = n if end == -1 else end + 3
            out.append(" ")
            continue

        if ch in {'"', "'", "`"}:
            quote = ch
            i += 1
            while i < n and src[i] != quote:
                if src[i] == "\\":
                    i += 1
                i += 1
            i += 1
            out.append(" ")
            continue

        out.append(ch)
        i += 1

    return "".join(out)


def fingerprint(source: str) -> str:
    """Identifiers → `v`, numbers → `n`, keywords and punctuation survive.

    Renaming variables or reformatting does not change the fingerprint;
    changing the algorithm does. That is exactly the property the Guard needs.
    """
    src = _strip_literals_and_comments(source)
    tokens: list[str] = []
    i = 0
    n = len(src)

    while i < n:
        ch = src[i]

        if ch.isspace():
            i += 1
            continue

        if _IDENT_START.match(ch):
            start = i
            while i < n and _IDENT_PART.match(src[i]):
                i += 1
            word = src[start:i].lower()
            tokens.append(word if word in KEYWORDS else "v")
            continue

        if ch.isdigit():
            while i < n and _NUMBER_PART.match(src[i]):
                i += 1
            tokens.append("n")
            continue

        tokens.append(ch)
        i += 1

    return " ".join(tokens)


def similarity(a: str, b: str) -> float:
    """Dice coefficient over token bigrams.

    Symmetric and length-tolerant — appropriate for "is this suspiciously close
    to the official answer?" rather than exact plagiarism detection.
    """

    def bigrams(s: str) -> Counter[str]:
        parts = [p for p in s.split(" ") if p]
        return Counter(f"{parts[i]} {parts[i + 1]}" for i in range(len(parts) - 1))

    ca, cb = bigrams(a), bigrams(b)
    total_a, total_b = sum(ca.values()), sum(cb.values())
    if total_a == 0 or total_b == 0:
        return 0.0
    overlap = sum(min(count, cb[key]) for key, count in ca.items())
    return (2 * overlap) / (total_a + total_b)

File: packages/db/src/fingerprint.ts
/**
 * Solution fingerprinting — TypeScript side.
 *
 * Produces an identifier-insensitive token stream from source code. The
 * Response Guard in the AI service compares a model's proposed code against
 * this fingerprint to detect that it is about to hand over the official
 * solution (docs 07 §5).
 *
 * IMPORTANT: `apps/ai/app/agents/fingerprint.py` implements the byte-identical
 * algorithm. The two are covered by a shared fixture test — if you change one,
 * change both, or the Guard silently stops catching leaks.
 */

const KEYWORDS = new Set([
  // control flow (shared across our 12 languages)
  'if', 'else', 'elif', 'for', 'while', 'do', 'switch', 'case', 'default',
  'break', 'continue', 'return', 'yield', 'goto',
  // declarations
  'def', 'function', 'fn', 'func', 'lambda', 'class', 'struct', 'interface',
  'enum', 'const', 'let', 'var', 'val', 'static', 'public', 'private',
  'protected', 'void', 'new', 'delete', 'this', 'self', 'super', 'extends',
  'implements', 'import', 'from', 'include', 'using', 'namespace', 'package',
  // types
  'int', 'long', 'short', 'char', 'float', 'double', 'bool', 'boolean',
  'string', 'str', 'list', 'dict', 'set', 'map', 'vector', 'array', 'auto',
  'unsigned', 'signed', 'size_t',
  // values / operators as words
  'true', 'false', 'null', 'none', 'nil', 'nullptr', 'undefined',
  'and', 'or', 'not', 'in', 'is', 'try', 'catch', 'except', 'finally',
  'throw', 'raise', 'with', 'as', 'pass', 'global', 'nonlocal',
  // very common library calls whose presence is structural, not cosmetic
  'range', 'len', 'enumerate', 'sorted', 'sort', 'append', 'push', 'pop',
  'min', 'max', 'sum', 'abs', 'print', 'push_back', 'insert', 'get',
]);

/** Strip comments and string literals without a real parser. */
function stripLiteralsAndComments(src: string): string {
  let out = '';
  let i = 0;
  const n = src.length;

  while (i < n) {
    const c = src[i]!;
    const next = i + 1 < n ? src[i + 1]! : '';

    // line comments: // and #
    if ((c === '/' && next === '/') || c === '#') {
      while (i < n && src[i] !== '\n') i++;
      continue;
    }
    // block comment: /* ... */
    if (c === '/' && next === '*') {
      i += 2;
      while (i + 1 < n && !(src[i] === '*' && src[i + 1] === '/')) i++;
      i += 2;
      continue;
    }
    // triple-quoted python strings
    if ((c === '"' || c === "'") && src.slice(i, i + 3) === c.repeat(3)) {
      const q = c.repeat(3);
      i += 3;
      const end = src.indexOf(q, i);
      i = end === -1 ? n : end + 3;
      out += ' ';
      continue;
    }
    // ordinary string / char literals
    if (c === '"' || c === "'" || c === '`') {
      const quote = c;
      i++;
      while (i < n && src[i] !== quote) {
        if (src[i] === '\\') i++;
        i++;
      }
      i++;
      out += ' ';
      continue;
    }
    out += c;
    i++;
  }
  return out;
}

const IDENT_START = /[A-Za-z_]/;
const IDENT_PART = /[A-Za-z0-9_]/;
const DIGIT = /[0-9]/;
const WHITESPACE = /\s/;

/**
 * Identifiers collapse to `v`, numeric literals to `n`, keywords survive
 * verbatim, and punctuation survives as single tokens. Renaming variables or
 * reformatting therefore does not change the fingerprint; changing the
 * algorithm does.
 */
export function fingerprint(source: string): string {
  const src = stripLiteralsAndComments(source);
  const tokens: string[] = [];
  let i = 0;

  while (i < src.length) {
    const c = src[i]!;

    if (WHITESPACE.test(c)) {
      i++;
      continue;
    }

    if (IDENT_START.test(c)) {
      let word = '';
      while (i < src.length && IDENT_PART.test(src[i]!)) {
        word += src[i]!;
        i++;
      }
      const lowered = word.toLowerCase();
      tokens.push(KEYWORDS.has(lowered) ? lowered : 'v');
      continue;
    }

    if (DIGIT.test(c)) {
      while (i < src.length && /[0-9.eExXaAbBcCdDfF]/.test(src[i]!)) i++;
      tokens.push('n');
      continue;
    }

    tokens.push(c);
    i++;
  }

  return tokens.join(' ');
}

/**
 * Dice coefficient over token bigrams. Symmetric, length-tolerant, and cheap —
 * appropriate for "is this suspiciously close to the official answer?" rather
 * than exact plagiarism detection.
 */
export function similarity(a: string, b: string): number {
  const bigrams = (s: string): Map<string, number> => {
    const parts = s.split(' ').filter(Boolean);
    const m = new Map<string, number>();
    for (let i = 0; i + 1 < parts.length; i++) {
      const key = `${parts[i]} ${parts[i + 1]}`;
      m.set(key, (m.get(key) ?? 0) + 1);
    }
    return m;
  };

  const ma = bigrams(a);
  const mb = bigrams(b);
  const totalA = [...ma.values()].reduce((x, y) => x + y, 0);
  const totalB = [...mb.values()].reduce((x, y) => x + y, 0);
  if (totalA === 0 || totalB === 0) return 0;

  let overlap = 0;
  for (const [k, v] of ma) overlap += Math.min(v, mb.get(k) ?? 0);
  return (2 * overlap) / (totalA + totalB);
}
